gluetool/glue.py

```diff
@@ -1978,57 +2066,81 @@ def run_module(self, module_name, module_argv=None, actual_module_name=None, reg
 
         return self.run_modules([step], register=register)
 
-    def module_list(self):
-        # type: () -> List[str]
+    def modules_as_groups(self, modules=None):
+        # type: (Optional[ModuleRegistryType]) -> Dict[str, ModuleRegistryType]
+        """
+        Gathers modules by their groups.
+
+        :rtype: dict(str, dict(str, DiscoveredModule))
+        :returns: dictonary where keys represent module groups, and values are mappings between
+            module names and the corresponding modules.
+        """
+
+        modules = modules or self.modules
+
+        groups = collections.defaultdict(dict)  # type: Dict[str, Dict[str, DiscoveredModule]]
+
+        for name, module_info in modules.iteritems():
+            groups[module_info.group][name] = module_info
+
+        return groups
+
+    def modules_descriptions(self, modules=None, groups=None):
+        # type: (Optional[ModuleRegistryType], Optional[List[str]]) -> str
+        """
+        Returns a string with modules and their descriptions.
+
+        :param dict(str, DiscoveredModule) modules: mapping with modules. If not set, current known modules
+            are used.
+        :param list(str) groups: if set, limit descriptions to modules belinging to the given groups.
+        :rtype: str
+        """
 
-        return sorted(self.modules)
+        modules = modules or self.modules
 
-    def module_list_usage(self, groups):
-        # type: (List[str]) -> str
+        as_groups = self.modules_as_groups(modules=modules)
 
-        """ Returns a string with modules description """
+        descriptions = []  # type: List[str]
 
         if groups:
-            usage = [
-                'Available modules in {} group(s)'.format(', '.join(groups))
+            descriptions += [
+                'Available modules in group(s) {}'.format(', '.join(groups))
             ]
+
         else:
-            usage = [
+            descriptions += [
                 'Available modules'
             ]
 
-        # get module list
-        plist = self.module_group_list()
-        if not plist:
-            usage.append('')
-            usage.append('  -- no modules found --')
+        def _add_no_modules():
+            # type: () -> None
+
+            descriptions.extend([
+                '',
```

Comment:
what as reason for this empty strings? New line in description?

Reply:
Yes. I will add a comment to `descriptions`.

Reply:
Should be fixed.